ledger/metrics/src/main/scala/com/daml/metrics/grpc/GrpcServerInterceptor.scala

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.metrics.grpc
+
+import com.daml.metrics.api.MetricHandle.Timer.TimerHandle
+import com.daml.metrics.api.MetricHandle.{Histogram, Meter, Timer}
+import com.daml.metrics.api.MetricsContext
+import com.daml.metrics.api.MetricsContext.{withExtraMetricLabels, withMetricLabels}
+import com.google.protobuf.{GeneratedMessage => JavaGeneratedMessage}
+import io.grpc.ForwardingServerCall.SimpleForwardingServerCall
+import io.grpc.ServerCall.Listener
+import io.grpc.{
+  ForwardingServerCallListener,
+  Metadata,
+  ServerCall,
+  ServerCallHandler,
+  ServerInterceptor,
+  Status,
+}
+import scalapb.{GeneratedMessage => ScalapbGeneratedMessage}
+
+class GrpcServerInterceptor(metrics: GrpcServerMetrics) extends ServerInterceptor {
+
+  override def interceptCall[ReqT, RespT](
+      serverCall: ServerCall[ReqT, RespT],
+      metadata: Metadata,
+      serverCallHandler: ServerCallHandler[ReqT, RespT],
+  ): ServerCall.Listener[ReqT] = {
+    val methodDescriptor = serverCall.getMethodDescriptor
+    val clientType = if (methodDescriptor.getType.clientSendsOneMessage()) "unary" else "streaming"
+    val serverType = if (methodDescriptor.getType.serverSendsOneMessage()) "unary" else "streaming"
+    withMetricLabels(
+      "grpc_service_name" -> methodDescriptor.getServiceName,
+      "grpc_method_name" -> methodDescriptor.getBareMethodName,
+      "grpc_client_type" -> clientType,
+      "grpc_server_type" -> serverType,
+    ) { implicit mc =>
+      val timerHandle = metrics.callTimer.startAsync()
+      val metricsServerCall = new MetricsCall(
+        serverCall,
+        timerHandle,
+        metrics.messagesSent,
+        metrics.messagesSentSize,
+        metrics.callsFinished,
+        metrics.messagesRequested,
+      )
+      new MonitoringServerCallListener(
+        serverCallHandler.startCall(metricsServerCall, metadata),
+        metrics.messagesReceived,
+        metrics.messagesReceivedSize,
+        metrics.callsStarted,
+      )
+    }
+  }
+
+  private final class MonitoringServerCallListener[T](
+      val delegate: Listener[T],
+      messagesReceived: Meter,
+      messagesReceivedSize: Histogram,
+      callsStarted: Meter,
+  )(implicit
+      metricsContext: MetricsContext
+  ) extends ForwardingServerCallListener[T] {
+
+    override def onMessage(message: T): Unit = {
+      super.onMessage(message)
+      updateHistogramWithSerializedSize(messagesReceivedSize, message)
+      messagesReceived.mark()
+    }
+    override def onReady(): Unit = {
+      super.onReady()
```

Comment:
iirc `onReady` is used with flow control so may be called several times for the same stream.

Reply:
Good catch, I moved it in the same place where we start the call.